store/file.go

```diff
@@ -0,0 +1,441 @@
+	metaFile, err := os.CreateTemp(filepath.Dir(metaPath), ".zstash-meta-*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp metadata file: %w", err)
+	}
+	tmpMeta := metaFile.Name()
+
+	cleanupMeta := true
+	defer func() {
+		_ = metaFile.Close()
+		if cleanupMeta {
+			_ = os.Remove(tmpMeta)
+		}
+	}()
+
+	if err := json.NewEncoder(metaFile).Encode(metadata); err != nil {
+		return nil, fmt.Errorf("failed to write metadata: %w", err)
+	}
+
+	if err := metaFile.Sync(); err != nil {
+		return nil, fmt.Errorf("failed to sync metadata file: %w", err)
+	}
+
+	if err := metaFile.Close(); err != nil {
+		return nil, fmt.Errorf("failed to close metadata file: %w", err)
+	}
+
+	// Remove existing metadata file before rename (required for Windows atomicity)
+	if err := os.Remove(metaPath); err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to remove existing metadata file: %w", err)
+	}
+
+	if err := os.Rename(tmpMeta, metaPath); err != nil {
+		return nil, fmt.Errorf("failed to rename metadata file: %w", err)
+	}
+
+	cleanupMeta = false
+
+	duration := time.Since(start)
+	averageSpeed := calculateTransferSpeedMBps(bytesWritten, duration)
+
+	span.SetAttributes(
+		attribute.Int64("bytes_transferred", bytesWritten),
+		attribute.String("transfer_speed", fmt.Sprintf("%.2fMB/s", averageSpeed)),
+		attribute.String("key", key),
+	)
+
+	return &TransferInfo{
+		BytesTransferred: bytesWritten,
+		TransferSpeed:    averageSpeed,
+		RequestID:        "",
+		Duration:         duration,
+	}, nil
+}
+
+// Download retrieves a cached file identified by key and writes it to destPath.
+//
+// The download process:
+//  1. Validates the cache key and destination path
+//  2. Reads the cached data file from storage
+//  3. Writes atomically to destination using temp file + fsync + rename
+//  4. Restores original file metadata (mtime) from sidecar if available (best-effort)
+//  5. Syncs parent directory for durability (best-effort)
+//
+// Metadata restoration is best-effort; failures are logged but don't fail the download.
+// Atomic writes ensure partial files are never visible at the destination path.
+//
+// Returns TransferInfo with bytes transferred, transfer speed, and duration.
+// Returns an error if the cache key doesn't exist or file operations fail.
+func (b *LocalFileBlob) Download(ctx context.Context, key string, destPath string) (*TransferInfo, error) {
+	_, span := trace.Start(ctx, "LocalFileBlob.Download")
+	defer span.End()
+
+	start := time.Now()
+
+	if err := validateFilePath(destPath); err != nil {
+		return nil, fmt.Errorf("invalid destination path: %w", err)
+	}
+
+	dataPath, metaPath, err := b.keyToPaths(key)
+	if err != nil {
+		return nil, err
+	}
+
+	srcFile, err := os.Open(dataPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open source file: %w", err)
+	}
+	defer func() {
+		_ = srcFile.Close()
+	}()
+
+	if err := os.MkdirAll(filepath.Dir(destPath), 0o755); err != nil {
```

Comment:
Is there a case where we'd be restoring to an existing path?

Reply:
@CerealBoy Yes `MkdirAll` works like `mkdir -p` so should be 👍🏻 whether the paths exist or not.

Was there another concern here?